### PR TITLE
Introduce //libs-scala/scala-utils

### PR DESCRIPTION
**Pull request by leo-da**
Moving `Statements.discard` from `//ledger-server/http-json` into `//libs-scala/scala-utils`.

`//libs-scala/scala-utils` is  a collection of generic scala utilities that most of the scala modules want to share.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: libs-scala/scala-utils/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+load(
+    "//bazel_tools:scala.bzl",
+    "da_scala_library",
+    "da_scala_test",
+    "lf_scalacopts",
+)
+
+hj_scalacopts = lf_scalacopts + [
```
Comment (cocreature) [inline libs-scala/scala-utils/BUILD.bazel:11]: What does hj stand for?

Comment (leo-da) [reply libs-scala/scala-utils/BUILD.bazel:11]: initially it was for Http Json I guess. I don't have a better name for it... suggestions?

If we use `"-P:wartremover:traverser:org.wartremover.warts.NonUnitStatements"` more widely, it makes sense to move it into `bazel_tools/scala.bzl` and place next to `lf_scalacopts` definition... but I am out of ideas what to call it :)

`lf_strict_scalaopts`???

Comment (cocreature) [reply libs-scala/scala-utils/BUILD.bazel:11]: Given that you are only using it in this file, maybe just call it `scalacopts` without a prefix?

**Comment by leo-da** [comment]
> Nice, thank you! `NonUnitStatements` looks very familiar to my haskell eyes

It also helps to prevent unhanded `Try[A]` and `Future[A]` errors. I agree with @S11001001 if you call a function, you want to do something with it's return value. If it has side-effects, just wrap the call in `discard { ... }` and it will be a conscious choice to disregard this specific value.

File: libs-scala/scala-utils/BUILD.bazel (hunk 1 above)
Comment (leo-da) [reply libs-scala/scala-utils/BUILD.bazel:11]: :+1:

File: libs-scala/scala-utils/src/main/scala/com/daml/scala/util/Statement.scala [hunk 2]
```diff
@@ -1,9 +1,16 @@
 // Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-package com.daml.http
+package com.daml.scala.util
```
Comment (S11001001) [inline libs-scala/scala-utils/src/main/scala/com/daml/scala/util/Statement.scala:4]: I suggest calling this `scalautil`. That's because if this is in classpath and you import `com.daml._`, you have a different `scala` in scope than the one you expect.

I hate to bikeshed, but we made this mistake elsewhere and it turned out to be more annoying than I expected, so we should probably avoid it here.

File: libs-scala/scala-utils/src/main/scala/com/daml/scalautil/Statement.scala [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
 object Statement {
+
+  /**
+    *  Suppresses `traverser:org.wartremover.warts.NonUnitStatements` warning on the expression level.
+    *
+    * @param evaluateForSideEffectOnly an expression with a side-effect that needs to be evaluated.
+    * @tparam A return type of the expression that gets discarded.
+    */
   @specialized def discard[A](evaluateForSideEffectOnly: A): Unit = {
     val _: A = evaluateForSideEffectOnly
```
Comment (SamirTalwar-DA) [inline libs-scala/scala-utils/src/main/scala/com/daml/scalautil/Statement.scala:15]: Why is this line necessary? `evaluateForSideEffectOnly` has already been evaluated at this point; it's not a thunk.

Comment (leo-da) [reply libs-scala/scala-utils/src/main/scala/com/daml/scalautil/Statement.scala:15]: I guess it is to avoid the compiler warning. If you do not assign it to `_`, you get:
```
libs-scala/scala-utils/src/main/scala/com/daml/scalautil/Statement.scala:14: warning: parameter value evaluateForSideEffectOnly in method discard is never used
  @specialized def discard[A](evaluateForSideEffectOnly: A): Unit = {
```

@S11001001 do you know if there is any other reason?

In any case the entire function has been "borrowed" from the `NonUnitStatements` wartremover docs:
http://www.wartremover.org/doc/warts.html

Comment (S11001001) [reply libs-scala/scala-utils/src/main/scala/com/daml/scalautil/Statement.scala:15]: No, that is the only reason.
